Expand docs for `julia` engine (#1573)
quarto-dev/quarto-web: docs/computations/julia.qmd

Quarto has two available engines for executing Julia code. The older one is using the [IJulia](https://github.com/JuliaLang/IJulia.jl) Jupyter kernel and depends on Python to run. The newer engine is using the [QuartoNotebookRunner.jl](https://github.com/PumasAI/QuartoNotebookRunner.jl/) package to render notebooks and does not have any additional dependencies beyond a Julia installation.

> To use either of these engines will require that you manually install Julia if
> you have not done so already. You can download it from

Reviewer: ```suggestion
Using either of these engines will require manually installing Julia if
you have not done so already. You can download it from
```

> https://julialang.org/downloads/. There are two options for installation: the
> Juliaup installation manager, and the "manual" installation approach. Unless
> you know that you need to use the "manual" approach you should prefer using
> Juliaup since it allows you to manage multiple Julia versions on your system.

Reviewer: ```suggestion
https://julialang.org/downloads/. There are two options for installation: the
Juliaup installation manager, or the "manual" installation approach. Unless
you know that you need to use the "manual" approach, use
Juliaup since it allows you to manage multiple Julia versions on your system.
```

- `exeflags`: An array of strings which are appended to the `julia` command that starts the worker process. For example, a notebook is run with `--project=@.` by default (the environment in the directory where the notebook is stored) but this could be overridden by setting `exeflags: ["--project=/some/directory/"]`.
- `env`: An array of strings where each string specifies one environment variable that is passed to the worker process. For example, `env: ["SOMEVAR=SOMEVALUE"]`.

> ### `juliaup` integration

Reviewer 2: This is cool ! Should `juliaup` be mentioned in the installation section: https://quarto.org/docs/computations/julia.html#installation-1 ?

Author: Probably best if it was.

[`juliaup`](https://github.com/JuliaLang/juliaup) is the recommended way to
install and manage Julia versions. The `julia` engine supports using
`juliaup`'s `+` channel specifier to select the Julia version that a notebook
uses. This allow users to run several notebooks concurrently with different
Julia versions on the same machine without the need for customizing their
`PATH` in any way.

> To use this feature ensure that you have used `juliaup` to install the channels

Reviewer 3: "feature, ensure" I think

> that you wish to use in your notebooks. The add the channel specifier to the notebook

Reviewer 3: "Then add"

frontmatter in the `julia.exeflags` option:

````markdown
---
title: "A Julia 1.11 notebook"
engine: julia
julia:
  exeflags: ["+1.11"]
---

```{{julia}}
VERSION
```
````

`QuartoNotebookRunner` currently supports running Julia versions from 1.6
onwards. Support for earlier versions is not planned.

### Revise.jl integration

[Revise](https://github.com/timholy/Revise.jl) allows for automatically
updating function definitions in Julia sessions. It is an essential tool in the
Julia ecosystem for any user that wishes to develop their own packages or large
projects. The `julia` engine supports using `Revise` within notebook processes
by simply importing the package into a cell at the start of a notebook. Once
imported `Revise` will automatically update functions imported from locally
developed packages in the same way as it behaves in a Julia REPL.

Ensure that `Revise` is installed in the project environment that the notebook
is using, since the global environment is not included in the load path
provided to Julia, unlike the behaviour of a Julia REPL session.

> ### Caching

Reviewer 2: Now that julia engine support caching, a link to this part from this part could be useful: https://quarto.org/docs/projects/code-execution.html#cache

BTW do `--cache-refresh` works with Julia engine ?

Author: > BTW do --cache-refresh works with Julia engine ?

Probably not currently. Do you happen to know if CLI flags get forwarded to the `julia` process along with all the frontmatter metadata from the notebook? If it does then we could probably support that behaviour.

Reviewer 2: I have quickly look 

`--cache-refresh` does set internal variable
https://github.com/quarto-dev/quarto-cli/blob/0502c32d72bbc90cdd4736135bd7f645e773ff37/src/command/render/flags.ts#L272-L275

that is use to modify `execute.cache` 
https://github.com/quarto-dev/quarto-cli/blob/0502c32d72bbc90cdd4736135bd7f645e773ff37/src/command/render/render-contexts.ts#L190-L193

so basically this is same as
```
execute:
  cache: refresh
```

So you have access to it already same way as Jupyter 
https://github.com/quarto-dev/quarto-cli/blob/0502c32d72bbc90cdd4736135bd7f645e773ff37/src/resources/jupyter/notebook.py#L116-L139

Does it help ?

Author: Yes, that does help, thanks.

frontmatter in the `julia.exeflags` option:

````markdown
---
title: "A Julia 1.11 notebook"
engine: julia
julia:
  exeflags: ["+1.11"]
---

```{{julia}}
VERSION
```
````

`QuartoNotebookRunner` currently supports running Julia versions from 1.6
onwards. Support for earlier versions is not planned.

### Revise.jl integration

[Revise](https://github.com/timholy/Revise.jl) allows for automatically
updating function definitions in Julia sessions. It is an essential tool in the
Julia ecosystem for any user that wishes to develop their own packages or large
projects. The `julia` engine supports using `Revise` within notebook processes
by simply importing the package into a cell at the start of a notebook. Once
imported `Revise` will automatically update functions imported from locally
developed packages in the same way as it behaves in a Julia REPL.

Ensure that `Revise` is installed in the project environment that the notebook
is using, since the global environment is not included in the load path
provided to Julia, unlike the behaviour of a Julia REPL session.

### Caching {#caching-julia}

The engine has built-in support for caching notebook results. This feature is disabled by

> default but can be enabled by setting `execute.cache` option to `true` in a notebook's

Reviewer: ```suggestion
default but can be enabled by setting the `execute.cache` option to `true` in a notebook's
```

frontmatter:

````markdown
---
title: "A caching example"
engine: julia
execute:
  cache: true
---

```{{julia}}
rand()
```
````

Notebook caches are invalidated based on the following criteria:

- Using a different Julia version to run the notebook.

> - The `Manifest.toml` of the environment the notebook is run in. Adding, removing, or changing package versions will invalidate the cache.

Reviewer: The first sentence is a fragment...maybe?
```suggestion
- Changes to the `Manifest.toml` of the environment the notebook is run in. Adding, removing, or changing package versions will invalidate the cache.
```
